File: thisweek_masuipeo2/tyama_codeiq2723.hs
#!/usr/bin/env runghc
import Data.List

-- A094047
fact i = if i<0 then 0 else foldl' (*) 1 [1..i]
comb n k = if k<0||n<k then 0 else foldl' (\r i -> (r*(n-i) `div` (i+1))) 1 [0..k-1]
func n = foldl' (\r i -> r + (-1)^i * (fact (n-i-1)) * (comb (2*n-i-1) i)) 0 [0..n-1]

main = readLn >>= \n -> print $ (-1)^n * 2 * (fact (n-1)) + (fact n) * (func n)

File: thisweek_masuipeo2/tyama_codeiq2723.rb
#!/usr/bin/ruby
#coding:utf-8

#A094047
def fact(n) (1..n).reduce(1,:*) end
def comb(n,k)
	r=1
	k.times{|i|r=r*(n-i)/(i+1)}
	r
end

N=gets.to_i
p (-1)**N * 2 * fact(N-1) + fact(N) * N.times.reduce(0){|s,i|s+(-1)**i * fact(N-i-1) * comb(2*N-i-1,i)}